apps/arweave/src/ar_wallet.erl

```diff
@@ -18,45 +17,86 @@
 -include_lib("eunit/include/eunit.hrl").
 -include_lib("public_key/include/public_key.hrl").
 
--define(PUBLIC_EXPNT, 65537).
-
 %% @doc Generate a new wallet public key and private key.
-new() ->
-	{[_, Pub], [_, Pub, Priv|_]} = {[_, Pub], [_, Pub, Priv|_]}
-		= crypto:generate_key(?SIGN_ALG, {?PRIV_KEY_SZ, ?PUBLIC_EXPNT}),
-	{{Priv, Pub}, Pub}.
+new() -> new(?DEFAULT_KEY_TYPE).
+new(KeyType = {KeyAlg, PublicExpnt}) when KeyType =:= ?DEFAULT_KEY_TYPE ->
+    {[_, Pub], [_, Pub, Priv|_]} = {[_, Pub], [_, Pub, Priv|_]}
+		= crypto:generate_key(KeyAlg, {?RSA_PRIV_KEY_SZ, PublicExpnt}),
+    {{KeyType, Priv, Pub}, {KeyType, Pub}};
+new(KeyType = {KeyAlg, KeyCrv}) when KeyAlg =:= ?ECDSA_SIGN_ALG andalso KeyCrv =:= secp256k1 ->
+    {Pub, Priv} = crypto:generate_key(ecdh, KeyCrv),
+    {{KeyType, Priv, Pub}, {KeyType, Pub}};
+new(KeyType = {KeyAlg, KeyCrv}) when KeyAlg =:= ?EDDSA_SIGN_ALG andalso KeyCrv =:= ed25519 ->
+    {Pub, Priv} = crypto:generate_key(KeyAlg, KeyCrv),
+    {{KeyType, Priv, Pub}, {KeyType, Pub}}.
 
 %% @doc Generate a new wallet public and private key, with a corresponding keyfile.
 new_keyfile() ->
-	new_keyfile(wallet_address).
+    new_keyfile(?DEFAULT_KEY_TYPE, wallet_address).
+new_keyfile(WalletName) ->
+    new_keyfile(?DEFAULT_KEY_TYPE, WalletName).
 
 %% @doc Generate a new wallet public and private key, with a corresponding keyfile.
 %% The provided key is used as part of the file name.
 %% @end
-new_keyfile(WalletName) ->
-	{[Expnt, Pub], [Expnt, Pub, Priv, P1, P2, E1, E2, C]} =
-		crypto:generate_key(rsa, {?PRIV_KEY_SZ, ?PUBLIC_EXPNT}),
-	Key =
-		ar_serialize:jsonify(
-			{
-				[
-					{kty, <<"RSA">>},
-					{ext, true},
-					{e, ar_util:encode(Expnt)},
-					{n, ar_util:encode(Pub)},
-					{d, ar_util:encode(Priv)},
-					{p, ar_util:encode(P1)},
-					{q, ar_util:encode(P2)},
-					{dp, ar_util:encode(E1)},
-					{dq, ar_util:encode(E2)},
-					{qi, ar_util:encode(C)}
-				]
-			}
-		),
+new_keyfile(KeyType, WalletName) ->
+    case KeyType of
+		{?RSA_SIGN_ALG, PublicExpnt} ->
+			{[Expnt, Pub], [Expnt, Pub, Priv, P1, P2, E1, E2, C]} =
+				crypto:generate_key(rsa, {?RSA_PRIV_KEY_SZ, PublicExpnt}),
+			Key =
+				ar_serialize:jsonify(
+					{
+						[
+							{kty, <<"RSA">>},
+							{ext, true},
+							{e, ar_util:encode(Expnt)},
+							{n, ar_util:encode(Pub)},
+							{d, ar_util:encode(Priv)},
+							{p, ar_util:encode(P1)},
+							{q, ar_util:encode(P2)},
+							{dp, ar_util:encode(E1)},
+							{dq, ar_util:encode(E2)},
+							{qi, ar_util:encode(C)}
+						]
+					}
+				);
+		{?ECDSA_SIGN_ALG, secp256k1} ->
+			{{_, Priv, Pub}, _} = new(KeyType),
+			<<_:8, PubPoint/binary>> = Pub,
+			PubPointMid = byte_size(PubPoint) div 2,
+			<<X:PubPointMid/binary, Y:PubPointMid/binary>> = PubPoint,
+			Key =
+				ar_serialize:jsonify(
+					{
+						[
+							{kty, <<"EC">>},
```

Comment:
Is it some common key storage format for ecdsa keys?

Reply:
Are you referring to the formatting of X and Y here? It appears to be. The Erlang OTP always generates public keys [uncompressed with an 0x04 prefix](https://bitcoin.stackexchange.com/questions/19666).

Here's an example of a [secp256k1 ECDSA jwk](https://identity.foundation/EcdsaSecp256k1RecoverySignature2020/#privateKeyJwk).

I think it would be good to add some tests loading externally generated keyfiles and signing/verifying with them.

Reply:
I thought about ECDSA public key compression a bit more and did some research.

It looks like Bitcoin supports both compressed and uncompressed EC public keys in its txs, see [here](https://en.bitcoin.it/wiki/Elliptic_Curve_Digital_Signature_Algorithm) and [here](https://bitcoin.stackexchange.com/questions/5952/why-does-bitcoin-support-both-compressed-and-uncompressed-keys-addresses). It will likely make more sense for us to only support compressed EC public keys in txs to save space and prevent confusion around the same EC wallets having two different addresses depending on client implementation.

Something that we can also take advantage of is the fact that EC public keys can be derived from their signatures. Ethereum appears to do this with their txs [here](https://cryptobook.nakov.com/digital-signatures/ecdsa-sign-verify-messages#ecdsa-public-key-recovery-from-signature). Not too sure on the trade-offs for implementing this since the changes to the node software might be substantial and this optimization is exclusive to ECDSA.

From what I can tell EdDSA signed txs don't require these considerations as its public keys are all already compressed.

I'll look into whether or not the Erlang OTP can verify ECDSA signatures with compressed public keys (it should be able to 🤷).

Reply:
Interesting, thank you. I wonder how fast the signature verification is compared to RSA, also considering compressed keys.

> Not too sure on the trade-offs for implementing this though since the changes to the node software might be substantial and this optimization is exclusive to ECDSA. It will also mess with the GraphQL API that lets you query for txs with the tx owner.

It does not require a protocol upgrade, we can introduce it later. Strategically though, I think, the more important work is including chain headers into the dataset for preservation and creating bundling services for scaling upload.

Reply:
I have updated the node to generate and load ECDSA public keys in a compressed form and made it so that it only accepts properly sized public keys for ECDSA and EdDSA.